File: backend/app/schemas/activity_log.py
from sqlmodel import SQLModel
import uuid
from datetime import datetime
from typing import Optional
from pydantic import model_validator

from .base import TunableBaseModel

# --- Schemi ActivityLog ---

class ActivityLogCreate(TunableBaseModel):
    start_time: datetime
    end_time: Optional[datetime] = None
    description: Optional[str] = None
    category_id: Optional[uuid.UUID] = None

    @model_validator(mode='after')
    def validate_end_time(self) -> 'ActivityLogCreate':
        if self.end_time and self.end_time <= self.start_time:
            raise ValueError("end_time must be after start_time")
        return self

class ActivityLogUpdate(TunableBaseModel):
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    description: Optional[str] = None
    category_id: Optional[uuid.UUID] = None

    # Nota: La validazione end_time > start_time è delegata al Service Layer
    # perché in un update parziale potremmo non avere entrambi i valori qui.

class ActivityLogRead(SQLModel):
    id: uuid.UUID
    start_time: datetime
    end_time: Optional[datetime]
    description: Optional[str]
    category_id: Optional[uuid.UUID]

File: frontend/lib/src/features/activity_log/data/models/activity_log.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'activity_log.freezed.dart';
part 'activity_log.g.dart';

@freezed
class ActivityLogCreate with _$ActivityLogCreate {
  const factory ActivityLogCreate({
    required DateTime startTime,
    DateTime? endTime,
    String? description,
    String? categoryId,
  }) = _ActivityLogCreate;

  factory ActivityLogCreate.fromJson(Map<String, dynamic> json) => _$ActivityLogCreateFromJson(json);
}

@freezed
class ActivityLogRead with _$ActivityLogRead {
  const factory ActivityLogRead({
    required String id,
    required DateTime startTime,
    DateTime? endTime,
    String? description,
    String? categoryId,
  }) = _ActivityLogRead;

  factory ActivityLogRead.fromJson(Map<String, dynamic> json) => _$ActivityLogReadFromJson(json);
}

@freezed
class ActivityLogUpdate with _$ActivityLogUpdate {
  const factory ActivityLogUpdate({
    DateTime? startTime,
    DateTime? endTime,
    String? description,
    String? categoryId,
  }) = _ActivityLogUpdate;

  factory ActivityLogUpdate.fromJson(Map<String, dynamic> json) => _$ActivityLogUpdateFromJson(json);
}
